File: ticket-manager-back/app/schemas/ticket.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import Category, Priority, TicketStatus


class TicketCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=5000)
    category: Category
    priority: Priority


class TicketOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    title: str
    description: str
    category: Category
    priority: Priority
    status: TicketStatus
    created_by_id: str
    created_at: datetime
    updated_at: datetime


class StatusHistoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    from_status: TicketStatus | None
    to_status: TicketStatus
    changed_by_name: str
    changed_at: datetime


class TicketDetail(BaseModel):
    id: str
    title: str
    description: str
    category: Category
    priority: Priority
    status: TicketStatus
    created_by_name: str
    created_at: datetime
    updated_at: datetime
    history: list[StatusHistoryOut]


class TicketStatusUpdate(BaseModel):
    status: TicketStatus
    note: str | None = Field(default=None, max_length=1000)


class TicketListItem(BaseModel):
    id: str
    title: str
    category: Category
    priority: Priority
    status: TicketStatus
    created_by_name: str
    created_at: datetime
    updated_at: datetime


class PaginatedTickets(BaseModel):
    items: list[TicketListItem]
    total: int
    page: int
    page_size: int

File: ticket-manager-front/src/types/ticket.ts
export type Category = "IT" | "FACILITIES" | "HR";
export type Priority = "LOW" | "MEDIUM" | "HIGH" | "URGENT";
export type TicketStatus = "OPEN" | "IN_PROGRESS" | "RESOLVED" | "CLOSED";

export const CATEGORY_LABELS: Record<Category, string> = {
  IT: "IT",
  FACILITIES: "Facilities",
  HR: "HR",
};

export const PRIORITY_LABELS: Record<Priority, string> = {
  LOW: "Low",
  MEDIUM: "Medium",
  HIGH: "High",
  URGENT: "Urgent",
};

export const STATUS_LABELS: Record<TicketStatus, string> = {
  OPEN: "Open",
  IN_PROGRESS: "In Progress",
  RESOLVED: "Resolved",
  CLOSED: "Closed",
};

export interface TicketCreatePayload {
  title: string;
  description: string;
  category: Category;
  priority: Priority;
}

export interface Ticket {
  id: string;
  title: string;
  description: string;
  category: Category;
  priority: Priority;
  status: TicketStatus;
  created_by_id: string;
  created_at: string;
  updated_at: string;
}

export interface TicketListItem {
  id: string;
  title: string;
  category: Category;
  priority: Priority;
  status: TicketStatus;
  created_by_name: string;
  created_at: string;
  updated_at: string;
}

export type SortBy = "created_at" | "priority";
export type SortDir = "asc" | "desc";

export interface TicketListFilters {
  status?: TicketStatus;
  category?: Category;
  priority?: Priority;
  sortBy: SortBy;
  sortDir: SortDir;
  page: number;
  // Opcional: por padrão usa PAGE_SIZE. A tela "Consultar / Atualizar Status"
  // sobrescreve com um valor menor (ver STATUS_PAGE_SIZE em TicketStatusPage.tsx).
  pageSize?: number;
}

export const DEFAULT_TICKET_FILTERS: TicketListFilters = {
  sortBy: "created_at",
  sortDir: "desc",
  page: 1,
};

export const PAGE_SIZE = 10;

export interface PaginatedTickets {
  items: TicketListItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface StatusHistoryItem {
  from_status: TicketStatus | null;
  to_status: TicketStatus;
  changed_by_name: string;
  changed_at: string;
}

export interface TicketDetail {
  id: string;
  title: string;
  description: string;
  category: Category;
  priority: Priority;
  status: TicketStatus;
  created_by_name: string;
  created_at: string;
  updated_at: string;
  history: StatusHistoryItem[];
}

// Espelha o mapa de transição sequencial do backend (ticket_service.ALLOWED_TRANSITIONS) —
// usado só pra saber qual é o próximo status a oferecer no botão, sem esperar a API responder.
export const ALLOWED_NEXT_STATUS: Record<TicketStatus, TicketStatus | null> = {
  OPEN: "IN_PROGRESS",
  IN_PROGRESS: "RESOLVED",
  RESOLVED: "CLOSED",
  CLOSED: null,
};
